File: platforms/android/keyboard-ui/src/main/java/app/funput/funput/keyboard/ui/emoji/EmojiSearchKeyboardLayout.kt
package app.funput.funput.keyboard.ui.emoji

import app.funput.funput.keyboard.model.KeyRole
import app.funput.funput.keyboard.model.KeySpec
import app.funput.funput.keyboard.model.KeyboardInputMethod
import app.funput.funput.keyboard.model.KeyboardLayout
import app.funput.funput.keyboard.model.KeyboardRow

internal object EmojiSearchKeyboardLayout {
    val layout = KeyboardLayout(
        id = "emoji-search-qwerty",
        inputMethod = KeyboardInputMethod.TELEX,
        suggestionBar = null,
        rows = listOf(
            letters("qwertyuiop"),
            letters("asdfghjkl", 0.5f),
            KeyboardRow(
                listOf(key("shift", "", KeyRole.SHIFT, 1.5f, "Shift")) +
                    "zxcvbnm".map(::letter) +
                    key("backspace", "", KeyRole.BACKSPACE, 1.5f, "Xóa"),
            ),
            KeyboardRow(
                listOf(
                    key("emoji", "☺", KeyRole.EMOJI, 1.7f, "Biểu tượng"),
                    key("space", "Tìm emoji", KeyRole.SPACE, 5.8f, "Dấu cách"),
                    key("done", "Xong", KeyRole.ENTER, 1.7f, "Xong"),
                ),
            ),
        ),
    )

    private fun letters(value: String, inset: Float = 0f) =
        KeyboardRow(value.map(::letter), inset)

    private fun letter(value: Char) = key(
        "search-$value",
        value.toString(),
        KeyRole.CHARACTER,
        shifted = value.uppercaseChar().toString(),
    )

    private fun key(
        id: String,
        label: String,
        role: KeyRole,
        weight: Float = 1f,
        accessibility: String = label,
        shifted: String? = null,
    ) = KeySpec(
        id,
        label,
        role,
        weight,
        shifted,
        accessibilityLabel = accessibility,
        spaceLabelOverride = label.takeIf { role == KeyRole.SPACE },
    )
}

File: platforms/ios/Packages/FunputKit/Sources/KeyboardRenderer/Emoji/Search/EmojiSearchKeyboardLayout.swift
#if canImport(UIKit)
import Foundation
import KeyboardLayout

enum EmojiSearchKeyboardLayout {
    static let layout = KeyboardLayout(
        id: "emoji-search-qwerty",
        inputMethod: .telex,
        toolbar: nil,
        rows: [
            row("qwertyuiop"),
            row("asdfghjkl", inset: 0.5),
            KeyboardRow(keys: [special("shift", .shift, weight: 1.5)]
                + keys("zxcvbnm")
                + [special("backspace", .backspace, weight: 1.5)]),
            KeyboardRow(keys: [
                special("emoji", .emoji, weight: 1.7),
                KeySpec(
                    id: "space", label: "Tìm emoji", role: .space,
                    widthWeight: 5.8, accessibilityLabel: "Dấu cách"
                ),
                special("done", .enter, weight: 1.7, label: "Xong"),
            ]),
        ]
    )

    private static func row(_ value: String, inset: CGFloat = 0) -> KeyboardRow {
        KeyboardRow(keys: keys(value), horizontalInsetUnits: inset)
    }

    private static func keys(_ value: String) -> [KeySpec] {
        value.map { character in
            let label = String(character)
            return KeySpec(
                id: "search-\(label)", label: label, role: .character,
                shiftedLabel: label.uppercased(), accessibilityLabel: label.uppercased()
            )
        }
    }

    private static func special(
        _ id: String, _ role: KeyRole, weight: CGFloat, label: String = ""
    ) -> KeySpec {
        KeySpec(
            id: "search-\(id)", label: label, role: role,
            widthWeight: weight, accessibilityLabel: label.isEmpty ? id : label
        )
    }
}
#endif
